### PR TITLE
[JENKINS-61687] Prevent concurrent build deletion

### PR DESCRIPTION
**Pull request by gmshake**
See [JENKINS-61687](https://issues.jenkins-ci.org/browse/JENKINS-61687).

### Proposed changelog entries

* ENKINS-61687: Prevent concurrent build deletion

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck
@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/model/Run.java [hunk 1]
```diff
@@ -1612,6 +1616,14 @@ public synchronized void deleteArtifacts() throws IOException {
      *      if we fail to delete.
      */
     public void delete() throws IOException {
+        synchronized (this) {
+            // Avoid concurrent delete.
+            if (state == State.PENDING_DELETE)
+                return;
```
Comment (stellargo) [inline core/src/main/java/hudson/model/Run.java:1622]:
```suggestion
            if (state == State.PENDING_DELETE) {
                return;
            }
```

File: core/src/main/java/hudson/model/Run.java [hunk 2]
```diff
@@ -1612,6 +1616,14 @@ public synchronized void deleteArtifacts() throws IOException {
      *      if we fail to delete.
      */
     public void delete() throws IOException {
+        synchronized (this) {
+            // Avoid concurrent delete.
```
Comment (stellargo) [inline core/src/main/java/hudson/model/Run.java:1620]:
```suggestion
            // Avoid concurrent delete. See https://issues.jenkins-ci.org/browse/JENKINS-61687
```

**Review by stellargo — APPROVED** [review]
Only slight nit pick. Implementation seems thread safe to me and looks like it should fix the bug referenced in Jira. But I have not tested it locally.

File: core/src/main/java/hudson/model/Run.java (hunk 1 above)
Comment (gmshake) [reply core/src/main/java/hudson/model/Run.java]: Fixed :)

File: core/src/main/java/hudson/model/Run.java (hunk 2 above)
Comment (gmshake) [reply core/src/main/java/hudson/model/Run.java]: Fixed :)

**Comment by gmshake** [comment]
Changed as @stellargo suggested, and rebased to lastest upstream master.

**Review by stellargo — APPROVED** [review]

**Comment by daniel-beck** [comment]
Thanks for your submission! This is an interesting approach.

I wonder whether it's possible to delete builds in progress, if so, this may mess things up for other uses of the `state`.

Still, this doesn't really add new API (worst case a new, immediately `@Deprecated` status), so it might be worth a try.

**Comment by gmshake** [comment]
A second approach would be that, we do not pollute `state` but introduce a boolean flag called `isPendingDelete`. The `pendingDelete` guards only the `delete()` method.

I'm not sure which one is better, but looks current logic in `delete()` does not consider builds those in progress.
So it might be problematical if a long running build would be deleted before it complete.

**Comment by daniel-beck** [comment]
Good point, that's probably safer to do and won't interfere with any public code handling `Run.State`, can be `private` or `protected`.

I wonder whether all major job implementations (Pipeline; Matrix; Maven, both module and parent; Freestyle) go through `Run#delete` though (i.e. at least calling the `super`). Something to check.

**Comment by gmshake** [comment]
Updated and use flag `isPendingDelete` . Local test looks good. :)

**Review by oleg-nenashev — COMMENTED** [review]
LGTM. In some cases it may to run deletion not happening due to error while API callers will get a "success" result. Maybe it makes sense to actually lock requests and propagate errors to all callers if an issue happens. Without async API and timeouts such implementation would be error-prone on its own, so maybe the fix is acceptable as a simple patch

**Comment by gmshake** [comment]
I aggree it's just a simple fix only preventing concurrent deleting.

The current state:
 current state

After the fix:
 after fix

**Comment by gmshake** [comment]
It looks more reasonable like this one, but requires introducing two new states : 
aaa

The `Run#delete()` need some state check like
```java
    /**
     * Deletes this build and its entire log
     *
     * @throws IOException
     *      if we fail to delete.
     */
    public void delete() throws IOException {
        synchronized (this) {
            if (state == State.DELETED) {
                return; // FIXME noop, or throw exception to indicate this build has been deleted ??
            }
            if (state == State.PENDING_DELETE) {
                return; // FIXME or throw exception to indicate concurrent deletion ??
            }
            if (state != State.COMPLETED) {
                // It looks not good to delete ongoing build.
                // FIXME or we cancel this build, then delete again ?
                throw new IllegalStateException("cannot delete build while in " + state);
            }

            state = State.PENDING_DELETE;
        }
        // It's safe to delete now.
        // Deleting process omitted.
        ...
       

        // Set state after deleting
        synchronized (this) {
            state = State.DELETED
        }
    }
```

**Review by olivergondza — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Thanks for the analysis @gmshake ! I think that the 2-state approach is a good solution here. It would have been an overkill to add states to the graph if `State` was a public API, but fortunately the `State` is a private class. I am +1 for the suggested solution, though it might require significant changes within the class to ensure the new states are properly interpreted everywhere. If you do not want to work on it right now, I think we could merge the current PR and create a follow-up task. From what I can tell, this PR is a net improvement anyway

**Review by timja — APPROVED** [review]

**Comment by gmshake** [comment]
@oleg-nenashev  Concurrency is hard. More work is required for the suggested solution.

> I think we could merge the current PR and create a follow-up task. 

I'm OK to merge the current PR :)

**Review by oleg-nenashev — APPROVED** [review]
I have created https://issues.jenkins-ci.org/browse/JENKINS-63275 as a follow-up.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Pull request merged**